### ISSUE TITLE
proposal: support bufferization in user code and provide method that can flush several events at once

### ISSUE THREAD

**Issue by rpokrovskij** [NLog/NLog#1792]
Situation: my user code bufferize "verbose" tracing messages till the system event (e.g. "interesting input" or exception ) and then flush them.
For better performance (as I imagine logging: for each logging operation there is expensive "open destination stream" operation ) it would be nice to have new `Logger` method like:
`        public void Log(IEnumerable<LogEventInfo> logEvents);`
that accepts several events. Now it is absent. (Do not worry about real time datetime, stacktrace and other attributes - If user would need them there is possibility to add them as properties)

There is of course possibility to log bufferized events at once as one big concatenated text, but this is not very attractive since there is strong feeling that bufferized event's record should no differ to much from not bufferized (some new custom properties that user can add by himself is OK)...

Thank you for your work.

**Comment by snakefoot** [comment]
You consider to wrap your target with the expensive-openstream in a [AsyncWrapper](https://github.com/NLog/NLog/wiki/AsyncWrapper-target) or [BufferingWrapper](https://github.com/NLog/NLog/wiki/BufferingWrapper-target)

**Comment by snakefoot** [comment]
Or you could vote on the idea to have flushtrigger, so you can send a special LogEvent that forces a flush of queued events. See #1712

**Comment by 304NotModified** [comment]
There is also https://github.com/NLog/NLog.ManualFlush

**Comment by rpokrovskij** [comment]
@snakefoot  nice possibilities, but I do not understand do you propose workarounds and support my proposal or do you say that NLOG really needs to force user to start use all those "superstructures" to get the performance boost in case of user implemented buffer?  
  
From my point of view: what can be simpler then just the base API method "push them all" which I have proposed? If you see any drawbacks - it would be nice to get an explanation.

**Comment by 304NotModified** [comment]
> It would be nice to have new Logger method like:
> public void Log(IEnumerable<LogEventInfo> logEvents);
You could also do `foreach(logevent in logevents){ logger.Log(logevent);}`? What's the benefit of the proposed method?

I'm not sure if I fully understand your proposal. Could you provide some more code calls/config examples?

**Comment by rpokrovskij** [comment]
@304NotModified   Yes, I use logging in loop right now, but I do a proposition that 100 calls in loops are expensive: they open a file (or any other target) 100 times, or locks it 100 times (when one can be enough for use case of flushing user buffer).

Nothing special in configuration I do not have in my minds. E.g. `<target name="verbose" type="File" fileName="..." />` and `<logger name="MyCategory:*" minlevel="Debug" writeTo="verbose" />`

**Comment by snakefoot** [comment]
Think you have to provide a PR with the code changes you want. There is buffering support for writing a batch of LogEvents, though by default it is only available through the AsyncTarget and BufferingTarget. This is why we recommend you to go in that official direction.

But you can pull the pants off NLog, and use the Logger to get the configured Target(s). For these targets then you have access to the method WriteAsyncLogEvents(). This can be used to write a batch of LogEvents.

**Comment by rpokrovskij** [comment]
@snakefoot PR? Just logic. Actually I'm not directly asking to provide "Log for batch" method but I'm asking to support custom bufferization. You see, user can buffer messages by himself better (more flexible)  than  with any common funcitionality that NLOG can provide. Do you confirm with that? Then what kind of support provides NLog for "custom buffering"? 

1. Messages can be logged in loop - drawback is the price of each Log call (manage resources 100 times for 100 calls).
2. Messages can be logged in the loop to BufferingTarget - drawback is overhead of NLog buferization. User had implemented custom buffer, then why NLOG pushes  him to use second "official" layer of buferization when bufferization is done already? 
3. WriteAsyncLogEvents to target find by name - from your description it should work. I will check this. But then if it so easy to usewhy just not to provide the same functioanlity  as Log(IEnumerable events) ?

**Comment by 304NotModified** [comment]
I not following it fully I think.

> Messages can be logged in loop - drawback is the price of each Log call (manage resources 100 times for 100 calls).

That's almost no overhead? When writing to a buffer, it's just a method call and add to a queue. AFAIK we do millions/sec. 

> @304NotModified Yes, I use logging in loop right now, but I do a proposition that 100 calls in a loop are expensive: they open a file (or any other target) 100 times, or locks it 100 times (when one can be enough for use case of flushing user buffer).

That's not the case if there is a bufferTarget between it. Events will be written in batches, so it won't open or lock 100 times. 

> why just not to provide the same functioanlity as Log(IEnumerable events) (of course if you agree that NLOG should support custom beffering)?

Because IMO the bufferTargetWrapper is needed for this case. I think the overhead is really to small to optimize for this case. But correct me if wrong.

**Comment by rpokrovskij** [comment]
> Because IMO the bufferTargetWrapper is needed for this case. I think the overhead is really to small to optimize for this case. But correct me if wrong.

Not in all cases. The last beta (or one of the last, I have tried it a week ago) was so slow (with log4net xml formatter) that I was scared and start thinking what can be improoved (but then just rollback to stable release).  So there are situations when overhead is not under control.   
   
If your IMO is based on using `WriteAsyncLogEvents`  and it will work for me (I will test it a little bit later) then I do not have arguments: NLOG supports custom bufferization (just not from public Logger interface). Nut if your IMO is based on "use internal buffer after custom" my argument is : you could be right saying "overhaed is small" but I'm right that NLOG doesn't support "custom bufferization".

**Comment by 304NotModified** [comment]
> Not in all cases. The last beta (or one of the last, I have tried it a week ago) was so slow (with log4net xml formatter) that I was scared and start thinking what can be improoved (but then just rollback to stable release).

While using the async or bufferwrapper? That should not be the case. (but I you like to check that, please open a new issue with details)

**Comment by rpokrovskij** [comment]
> But you can pull the pants off NLog, and use the LogManager.LoggingConfiguration to get the wanted target (FindTargetByName). For the wanted target then you have access to the method WriteAsyncLogEvents(). This can be used to write a batch of LogEvents.

WriteAsyncLogEvents logs "special" events `AsyncLogEventInfo` . Trying to convert my batch of LogEventInfo to them with 

`new AsyncLogEventInfo(logEventInfo,(ex)=> { if (ex!=null) throw ex; });`

I'm not sure that throwing exception in continuation is a good decision, but in any case my tests (file, log4net formater) improves  x2 times, it is a lot, and I like it.

But it seems I should hardcode/"configure in code" targetname additionally to "logger name".
We can say that NLOG supports "bufferization in custom code" only for Targets but not for Loggers. And this looks for me "inconsistent".

**Comment by snakefoot** [comment]
The exception is not for you to throw, but you to log. It comes from the  Target, when it has problems with logging.

**Comment by rpokrovskij** [comment]
@snakefoot,  thank you. It works. I have only minor question:
In which cases user would not want to support  LogManager.ThrownExceptions? It should be recommended and default behavior...

I see that "batch logging" and "async logging" are twisted together by NLOG design (Write**Async**LogEvent**s**). It explains why there are no batch logging on Logger interface: because Logger hides "assync, or not assync" specific and this is absolutely correct behavior.  But batch logging is not async logging! They are two independent concepts! So I should underline when I'm asking to add Batch Logging to Logger, I'm not asking to add "async batch logger" but just possibility to log several LogEventInfo in one logger operation.

**Comment by snakefoot** [comment]
> In which cases user would not want to support LogManager.ThrownExceptions?

The official NLog recommendation is to have LogManager.ThrownExceptions = false in production.

**Comment by rpokrovskij** [comment]
@snakefoot  can' understand your note. of course logging should be silent, and **I expect**  that `LogManager.ThrownExceptions = false `  will discard and also discard my thrown new `NLogRuntimeException("Exception occurred in NLog", ex)` in continuation. But that not means that NLOG recommendation is not use ThrownExceptions "bus".

**Comment by 304NotModified** [comment]
"Bus" ?

**Comment by 304NotModified** [comment]
> Not in all cases. The last beta (or one of the last, I have tried it a week ago) was so slow (with file target and log4net xml formatter) that I was scared (x500 times slower then current stable) 

Is that NLog 5 beta3? It should be not 500x times slower! Could you provide a testprogram for that? 

>  you could be right saying "overhaed is small" but I'm right that NLOG doesn't support "custom bufferization".

We don't support it indeed. And if there is no big difference in performance, I don't see why we should support it. We like to support stuff when I has benefits. Dropping a `for` loop isn't a great benefit, but it will make NLog's code more complex. 


> 1. Messages can be logged in loop
> 2. Messages can be logged in the loop to BufferingTarget 
> 3. WriteAsyncLogEvents to target find by name

> I'm not sure that throwing exception in continuation is a good decision, but what is important my tests (file, log4net formater, no wrappers) improves x2 times, it is a lot, and I like it.

I can't image, if 2 is well configured, that 3 will be 2 times faster than 3! Could you provide a testprogram for that? 

Please keep in mind that 3 is a work around, and you're skipping some NLog core features, like routing, error handling & filtering. 

> Final verdict: NLOG supports "bufferization in custom code" only for Target but not for Logger. It works for me (in my configuration "verbose" logger is common and has only one target) but configurations can be much more complex and in any case it looks "inconsistent" from API design.

We only support writing one event at the time, not batched writing. The API is consistent for that. Calling `WriteAsyncLogEvents ` is more a work around. 

> I see that "batch logging" and "async logging" are twisted together by NLOG design (WriteAsyncLogEvents). It explains why there are no batch logging on Logger interface: because Logger hides "assync, or not assync" specific (and this is absolutely correct behavior) together hiding from user "batch logging" methods. But batch logging is not async logging! They are two independent concepts!

Yes we are aware of that. But it isn't practical to combine async without batched logging. 

To conclude, only if there is a great performance (+25% avg cases) benefit of batched log writing, I would consider implementing '`logger.Log(IEnumerable<LogEventInfo>events)` - we need a testprogram to verify that.

**Comment by rpokrovskij** [comment]
> Is that NLog 5 beta3? It should be not 500x times slower! Could you provide a testprogram for that?
By date it was beta3.  Unfortunatly I can't repeat it right now. 
Not 500x, but 300x.  2 min against normal 400 px.  Those numbers are "whole test case" (dozen db operation each generate 60 event buferrized in custom buffer and then flushed to the file).
I was sure that this was NLOG problem since when I comment out of Logger.Log in loop , then performance returned to normal 400px
Again, unfortunatly I can't repeat it right now (may be it depends on runtime environment, may be because I had improoved verbose logging and now log less messages)  But I have  left beta3 in usage, and will inform you in case of problem returns.

Configuration used:
```
<nlog internalLogToTrace="true">
    <targets>
      <target name="authentication" type="File" fileName="${basedir}/App_Data/authentication.log.txt" layout="${date:format=dd.MM.yy HH\:mm\:ss}; ${event-context:item=Title}; ${message}" />
      <target name="activities" type="File" fileName="${basedir}/App_Data/activities.log.txt" layout="${date:format=dd.MM.yy HH\:mm\:ss}; ${event-context:item=Title}; ${message}" />
      <target name="verbose" type="File" fileName="${basedir}/App_Data/verbose.log.xml" layout="${log4jxmlevent:includeSourceInfo=false:includeCallSite=true:includeMDC=true:appInfo=true:includeNDC=false:includeNLogData=true:indentXml=true}" />
    </targets>
    <rules>
      <logger name="Routine:*" level="Trace"    writeTo="activities" />
      <logger name="Routine:*" minlevel="Debug" writeTo="verbose" />
      <logger name="Authentication" minlevel="Debug" writeTo="authentication" />
    </rules>
  </nlog>    

```

> I can't image, if 2 is well configured, that 3 will be 2 times faster than 3! Could you provide a testprogram for that?

What is well configured? I do not use NLOG target buffers. Is it incorrect? I will use them if you will push me, but naturally I would like to not have additional buffers after my (may be it is not naturally for you, but my application works in heavy magnetic field environment and hardware hangs quite often).

If you not expect that "well configured" is with target buffer I can create testprogram for this.

**Comment by 304NotModified** [comment]
> What is well configured? I do not use NLOG target buffers. Is it incorrect? I will use them if you will push me, but naturally I would like to not have additional buffers after my (may be it is not naturally for you, but my application works in heavy magnetic field environment and hardware hangs quite often).

I mean't, if a BufferWrapper (option 2) was used, then the performance compared to the `WriteAsyncLogEvents` (option 3) should not be 25%. That option 1 gives performance problems in your case, is conform expectations (from my side). 

> If you not expect that "well configured" is with target buffer I can create testprogram for this.

If there is a performance difference between option 2 and 3, then please provide a testprogram. If it's between 1 and 3, please try option 2.

**Comment by rpokrovskij** [comment]
> If there is a performance difference between option 2 and 3, then please provide a testprogram. If it's between 1 and 3, please try option 2.

I have just compared BufferingWrapper (2) and WriteAsyncLogEvents (3) and have found that (3) is two times faster. 

Of course it is because default BufferingWrapper  configuration (bufferSize 100) is not relevant for my case.
And I'm absolutely sure that you can propose a BufferingWrapper configuration that will loose less then 25% in speed against 3 but what that can prove?  Why you still thinking that it is a better idea to push user to think about how to configure NLog buffer then provide possibility to avoid using it?

**Comment by 304NotModified** [comment]
the necessity for this feature

**Comment by 304NotModified** [comment]
Thanks for this idea and elaborating it @rpokrovskij ! 

We don't think this feature we have to much impact and hence it will be a "wontfix"

**Comment by rpokrovskij** [comment]
> the necessity for this feature

It is not fair. For each testprog you can adjust NLOG bufferization and then say "Hey. it is not necessary".  But real application is not testprog and anybody else can't do the same easy as you.

Lets do it fairy. I describe test case, and you propose BufferingWrapper configuration (may be several). I will test it and see if your configurations fails against WriteAsyncLogEvents I will send it to you. Then it will be "prove"

Test case: looping "loops of events" (batches for 10-50 events), target: file, formatter log4j, each message 0-4KB.

**Comment by 304NotModified** [comment]
Please understand that changing the Logger interface and NLog core, have a large impact! We're trying to avoid that when it isn't absolute necessary.

If you send a testprogram (or github repos), I would be happy to look at it.

**Comment by rpokrovskij** [comment]
> Please understand that changing the Logger interface and NLog core, have a large impact! We're trying to avoid that when it isn't absolute necessary.

Don't worry. This I can understand.  Again WriteAsyncLogEvents works for me.

> If you send a testprogram (or github repos), I would be happy to look at it.

Can't understand this. Of course for each test case you will be able to adjust BufferingWrapper to show good performance (-25%). Why you will be happy... But may be to see why you will be happy I will send it to you.

**Comment by 304NotModified** [comment]
> But what will make you happy.

Tried to say: I will look at it

**Comment by rpokrovskij** [comment]
@304NotModified How I can send you a test prog (not expirienced githubuser)

Created as it was described. BufferSize="100000" doesn't help a lot.

Results: 
in loop + buffered  = 4 sec. 
WriteAsyncLogEvents  = 394 ms.

NLog used - betta 3

![tests](https://cloud.githubusercontent.com/assets/11598038/20869253/1bad766a-ba77-11e6-90fc-3e7e9d5b51c1.PNG)

[NLogTest.zip](https://github.com/NLog/NLog/files/629823/NLogTest.zip)

**Comment by rpokrovskij** [comment]
I've found an error in my test comparision of buffered against WriteAsyncLogEvents
 Should agree buffered is quicker.

**Comment by rpokrovskij** [comment]
BTW, 
1) WriteAsyncLogEvents + log4j formatter doesn't serialize logEventInfo custom properties.

2) In tests with parallel executions BufferingWrapper always loose 25% but then AsyncWrapper can show the equal results.  So I will possibly stick with AsyncWrapper, just need to test it better.

[NLogTest.zip](https://github.com/NLog/NLog/files/631141/NLogTest.zip)

![testa](https://cloud.githubusercontent.com/assets/11598038/20886162/5a2abda4-bafd-11e6-9303-f183e1b7490f.PNG)

**Comment by 304NotModified** [comment]
Thanks for testing!



> In tests with parallel executions BufferingWrapper always loose 25%

That is probably because of the flushTimeout="2000". You could flush it manual: `LogManager.Flush()`. Could you try that?

**Comment by rpokrovskij** [comment]
@304NotModified I can't understand how Flash can help in parallel executions. BufferingWrapper now shows 17 sec .

![testa](https://cloud.githubusercontent.com/assets/11598038/20892624/a3609d62-bb17-11e6-9ddf-df0ceba9a91c.PNG)

 ```
 <nlog internalLogToTrace="true" throwExceptions="true">
    <targets>
      <target type="BufferingWrapper"
              slidingTimeout="False"
              bufferSize="10000"
              name="VerboseBufferingWrapper">
        <target name="verbose1" type="File" fileName="${basedir}/App_Data/verbose1.log.xml" layout="${log4jxmlevent:includeSourceInfo=false:includeCallSite=true:includeMDC=true:appInfo=true:includeNDC=false:includeNLogData=true:indentXml=true}" />
      </target>
      <target type="AsyncWrapper"
              queueLimit="10000"
              timeToSleepBetweenBatches="50"
              batchSize="100"
              overflowAction="Discard"
              name="VerboseAsyncWrapper">
        <target name="verbose2" type="File" fileName="${basedir}/App_Data/verbose2.log.xml" layout="${log4jxmlevent:includeSourceInfo=false:includeCallSite=true:includeMDC=true:appInfo=true:includeNDC=false:includeNLogData=true:indentXml=true}" />
      </target>
      <target name="verbose3" type="File" fileName="${basedir}/App_Data/verbose3.log.xml" layout="${log4jxmlevent:includeSourceInfo=false:includeCallSite=true:includeMDC=true:appInfo=true:includeNDC=false:includeNLogData=true:indentXml=true}" />
    </targets>
    <rules>
      <logger name="Routine1:*" minlevel="Debug" writeTo="VerboseBufferingWrapper" />
      <logger name="Routine2:*" minlevel="Debug" writeTo="VerboseAsyncWrapper" />
      <logger name="Routine3:*" minlevel="Debug" writeTo="verbose3" />
    </rules>
  </nlog>
```

[NLogTest.zip](https://github.com/NLog/NLog/files/631635/NLogTest.zip)

Thank you for reopening. Of course I should say that test is not very realistic just because in my case only one user really would use "custom bufferization". And with one user "who log in batches" NLOG standard buffers log with same results as WriteAsyncLogEvents. I'm just trying to say that support custom bufferization can be more important then e.g. "flush on event" which you choose. Why I use custom buffer? Because I bufferize events on  user based level that means I  enable "verbose log" for concrete user or even concreate action that was marked as very "slow" in past... Then my code enables "verbose" loggin. collect all info and send it to me. NLOG will never provide such functionality.

**Comment by snakefoot** [comment]
> for all those "log only on exception, otherwise discard" tricks

If you are only interested in logging Exceptions with stacktrace, then you should just use LogLevel.Error for exceptions, and for all the other stuff you are logging that you don't want, then you can use LogLevel.Info.

**Comment by rpokrovskij** [comment]
@snakefoot my opinion  #1710 contributors do not understand fully how "flush on exception otherwise discard" functionality is mostly useful. Particullary I can't understand how it would work in multiuser environment. What I want: NLOG should support custom bufferization. What I not recommend: implement the scenario "flush on exception otherwise discard" internally. NLOG could just say to its users "if you want custom buffers with tricky scenarios, go and create them by yourself, there is a way".

My custom buffers are "user based". It is is a multiuser environment and when a user starts an action I check my own configuration at runtime (on parameters "action name" , "user profile", "input") and choose the logging scenario, one of scenarios is: log verbose messages to the buffer (only this user messages) and flush the buffer on condition (condition can be e.g. "action is too slow").

I do not want NLOG implement this. I have this already. But I'm asking to better support of custom buffers in NLOG (Logger.Log(IEnumerable)).

**Comment by snakefoot** [comment]
I could understand if you wanted some kind of batch-id (but that could just be a correlation-id on all the log-events for one user like ex the userid).

But there is little performance difference between adding them one at a time to the AsyncWrapper-Queue, or adding them in a single operation. There will be an iteration of the LogEvent-ienumerable, just a matter of where.

**Comment by rpokrovskij** [comment]
@snakefoot I do not want anything from NLOG except "Log(IEnumerable)" that means support to custom buffers. Also I worry you about implementing reach buffers ( #1710 ) inside the NLOG. It will be always not enough  flexible.

**Comment by 304NotModified** [comment]
> I do not want anything from NLOG except "Log(IEnumerable)"

Indeed, but because this has large impact, we need a very good reason. 

> and API is not so consistent how it could be.

it's consistent as every log operation is 1 event (logger). There is no need to be consistent over loggers and targets IMO.

> But again it use more other resources

I'm doubting if this a theoretical problem.  Will have to look to your test program though.

**Comment by rpokrovskij** [comment]
> I'm doubting if this a theoretical problem. 

Resourcses are not a problem at all: we waste them till lack of resource become a problem.
The theoretical problem is: should the NLOG implements "fancy bufferization tricks" by itself or just support custom buffer... But there is nothing about this in my testprog. Testprog just compare three ways of bufferization that NLOG provides in parallel execution conditions. 

> Indeed, but because this has large impact, we need a very good reason.

It is true, I can imagine... But from other hand - its clear that this is possible, you have what custom buffers needs on Target layer already.

**Comment by snakefoot** [comment]
#1853 (NLog ver. 4.4.1) should have improved the performance of the BufferingWrapper, so it should be faster for manual buffering and flushing.

**Comment by 304NotModified** [comment]
Thanks for the up @snakefoot 

Still thinking this would be nice, but not nice enough to change the core of NLog - it's have to much impact and the low benefits (IMO). So closing this one. 

Thanks for the discussion @rpokrovskij !
